### PR TITLE
Add `refit` parameter into `backtest`

### LINKED ISSUE #1154: Add `refit` parameter into `backtest`
### 🚀 Feature Request

After adding possibility to forecast on future with gap, we can modify our backtest to work with this feature.

### Proposal

Add parameter `refit` into `BasePipeline.backtest`.
 
Possible values:
- `True` (default)
  - fit on every fold
  - all folds should be processed in parallel (like now)
- `False`
  - fit only on the earliest fold
  - first fold is processed separately and pipeline is fit
  - consecutive folds should be processed in parallel
- `value: int`
  - refit pipeline every `value` folds starting from the earliest
  - in every group of folds that uses the same pipeline we should work like with `refit=False`

How to make this change:
- Extract logic of parallel run of folds from `BasePipeline.backtest` into a separate method
- Modify `BasePipeline._run_fold` to return fitted pipeline and accept fitted pipeline
- Implement logic with fitting folds and making predictions
  - We can interpret `refit=True` as `refit: int = 1`, because we consider each fold to be in the separate group
  - We can interpret `refit=False` as `refit: int = n_folds`, because we consider all folds to be in the same group

To discuss: what to do with `FoldMask`?

### Test cases

* Fix tests for changed methods
* Check that `n_jobs` doesn't influence the results with different `refit`
* Check that pipeline is working with models and transforms that supports inference mode
* Check that pipeline doesn't work with models and transforms that doesn't support inference mode

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Mr-Geekman**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Proposed Changes
Look #1154.

## Closing issues
Closes #1154.

### CODE REVIEW

File: etna/pipeline/base.py [hunk 1]
```diff
@@ -620,6 +654,114 @@ def _prepare_fold_masks(self, ts: TSDataset, masks: Union[int, List[FoldMask]],
...
+    def _make_backtest_fold_groups(masks: List[FoldMask], refit: Union[bool, int]) -> List[FoldParallelGroup]:
+        """Make groups of folds for backtest."""
+        if not refit:
+            refit = len(masks)
+
+        grouped_folds = []
+        num_groups = math.ceil(len(masks) / refit)
+        for group_id in range(num_groups):
+            train_fold_number = group_id * refit
+            forecast_fold_numbers = [train_fold_number + i for i in range(refit) if train_fold_number + i < len(masks)]
```
Comment (alex-hse-repository) [inline etna/pipeline/base.py:667]: Shouldn't it be `range(1, refit+1)`? Now we have the first fold common for train and forecast

File: etna/pipeline/base.py [hunk 2]
```diff
@@ -620,6 +654,114 @@ def _prepare_fold_masks(self, ts: TSDataset, masks: Union[int, List[FoldMask]],
...
+                        pipeline=pipeline,
+                        forecast_params=forecast_params,
+                    )
+                    for forecast_ts in group_forecast_datasets
+                )
+                for pipeline, group_forecast_datasets in zip(pipelines, forecast_datasets)
+            ]
+
+            # process forecasts
+            fold_process_train_datasets = (
```
Comment (alex-hse-repository) [inline etna/pipeline/base.py:729]: Looks like it is the same as `fit_datasets `

File: etna/pipeline/base.py [hunk 3]
```diff
@@ -620,6 +654,114 @@ def _prepare_fold_masks(self, ts: TSDataset, masks: Union[int, List[FoldMask]],
...
+                train for train, _ in self._generate_folds_datasets(ts=ts, masks=fit_masks, horizon=self.horizon)
+            )
+            fold_process_test_datasets = (
```
Comment (alex-hse-repository) [inline etna/pipeline/base.py:732]: The same here

File: tests/test_pipeline/test_pipeline.py [hunk 4]
```diff
@@ -439,15 +441,82 @@ def test_get_fold_info_interface_hours(catboost_pipeline: Pipeline, example_tsdf
     assert expected_columns == sorted(info_df.columns)
 
 
+def test_get_fold_info_refit_true(catboost_pipeline: Pipeline, example_tsdf: TSDataset):
+    """Check that Pipeline.backtest returns info dataframe with correct train with regular refit."""
+    _, _, info_df = catboost_pipeline.backtest(
```
Comment (alex-hse-repository) [inline tests/test_pipeline/test_pipeline.py:446]: May be we can use light pipeline with Naive model or sth like this for this test?

File: tests/test_pipeline/test_pipeline.py [hunk 5]
```diff
@@ -439,15 +441,82 @@ def test_get_fold_info_interface_hours(catboost_pipeline: Pipeline, example_tsdf
...
+        ts=example_tsdf, n_jobs=1, metrics=DEFAULT_METRICS, n_folds=5, refit=True
+    )
+    assert info_df["train_start_time"].nunique() == 1
```
Comment (alex-hse-repository) [inline tests/test_pipeline/test_pipeline.py:449]: Don't we want to check the exact values everywhere?

File: tests/test_pipeline/test_pipeline.py [hunk 6]
```diff
@@ -439,15 +441,82 @@ def test_get_fold_info_interface_hours(catboost_pipeline: Pipeline, example_tsdf
...
+    )
+    expected_refits = math.ceil(n_folds / refit)
+    assert info_df["train_start_time"].nunique() == 1
+    assert info_df["train_end_time"].nunique() == expected_refits
+    assert info_df["test_start_time"].nunique() == len(info_df)
+    assert info_df["test_end_time"].nunique() == len(info_df)
+
+
+def test_backtest_refit_success(catboost_pipeline: Pipeline, big_example_tsdf: TSDataset):
+    """Check that backtest with rare refit works on pipeline that supports it."""
```
Comment (alex-hse-repository) [inline tests/test_pipeline/test_pipeline.py:495]: With rare -> without??

File: tests/test_pipeline/test_pipeline.py [hunk 7]
```diff
@@ -439,15 +441,82 @@ def test_get_fold_info_interface_hours(catboost_pipeline: Pipeline, example_tsdf
...
+
+
+def test_backtest_refit_fail(big_example_tsdf: TSDataset):
+    """Check that backtest with rare refit doesn't work on pipeline that doesn't support it."""
+    pipeline = Pipeline(
+        model=NaiveModel(lag=7),
+        transforms=[DifferencingTransform(in_column="target", inplace=True)],
+        horizon=7,
+    )
+    with pytest.raises(ValueError, match="Test should go after the train without gaps"):
```
Comment (alex-hse-repository) [inline tests/test_pipeline/test_pipeline.py:506]: May be add some notes about this behaviour in the docs to backtest? It might be not obvious what happned

File: tests/test_pipeline/test_pipeline.py [hunk 8]
```diff
@@ -564,17 +633,103 @@ def test_generate_folds_datasets_without_first_date(ts_name, mask, request):
...
-def test_run_fold(ts_run_fold: TSDataset, mask: FoldMask, expected: Dict[str, List[float]]):
-    train, test = ts_run_fold.train_test_split(
+def test_process_fold_forecast(ts_process_fold_forecast, mask: FoldMask, expected: Dict[str, List[float]]):
+    train, test = ts_process_fold_forecast.train_test_split(
         train_start=mask.first_train_timestamp, train_end=mask.last_train_timestamp
     )
 
     pipeline = Pipeline(model=NaiveModel(lag=5), transforms=[], horizon=4)
-    fold = pipeline._run_fold(train, test, 1, mask, [MAE()], forecast_params=dict())
+    forecast = pipeline.forecast(ts=train)
```
Comment (alex-hse-repository) [inline tests/test_pipeline/test_pipeline.py:642]: Wow, forecast without fitting!))))

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: etna/pipeline/base.py (hunk 2 above)
Comment (Mr-Geekman) [reply etna/pipeline/base.py:729]: I can't reuse `fit_datasets`, it becomes empty after one iteration.

File: etna/pipeline/base.py (hunk 3 above)
Comment (Mr-Geekman) [reply etna/pipeline/base.py:732]: It is different, I use `test` here instead of `train`.

File: etna/pipeline/base.py (hunk 1 above)
Comment (Mr-Geekman) [reply etna/pipeline/base.py:667]: Currently `range(1, refit+1)` isn't working. I expect this numbers to show index in original `masks` list.

File: tests/test_pipeline/test_pipeline.py [hunk 9]
```diff
@@ -564,17 +633,103 @@ def test_generate_folds_datasets_without_first_date(ts_name, mask, request):
...
+
+def test_make_backtest_fold_groups_refit_false():
+    masks = [MagicMock() for _ in range(2)]
+    obtained_results = Pipeline._make_backtest_fold_groups(masks=masks, refit=False)
+    expected_results = [
+        {
+            "train_fold_number": 0,
+            "train_mask": masks[0],
+            "forecast_fold_numbers": [0, 1],
+            "forecast_masks": [masks[0], masks[1]],
```
[the remark this answers is not in the document]
Comment (Mr-Geekman) [reply tests/test_pipeline/test_pipeline.py:678]: Because we have to forecast the first fold of the group.

File: etna/pipeline/base.py [hunk 10]
```diff
@@ -229,6 +232,15 @@ def backtest(
             If True aggregate metrics above folds, return raw metrics otherwise
         n_jobs:
             Number of jobs to run in parallel
+        refit:
+            Determines how often pipeline should be retrained during iteration over folds.
+
+            * If ``True``: pipeline is retrained on each fold.
+
+            * If ``False``: pipeline is trained on the first fold, and on future folds trained pipeline is used.
```
[the remark this answers is not in the document]
Comment (alex-hse-repository) [reply etna/pipeline/base.py:240]: I mean without the rest part))

File: etna/pipeline/base.py [hunk 11]
```diff
@@ -656,6 +808,11 @@ def backtest(
         -------
         metrics_df, forecast_df, fold_info_df: Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]
             Metrics dataframe, forecast dataframe and dataframe with information about folds
+
+        Raises
+        ------
+        Exception:
```
Comment (alex-hse-repository) [inline etna/pipeline/base.py:814]: Not sure that this is really informative)

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: etna/pipeline/base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Mr-Geekman) [reply etna/pipeline/base.py:240]: Will be fixed on a couple of minutes

**Review by alex-hse-repository — APPROVED** [review]
